crates/hir_def/src/type_ref.rs

```diff
@@ -290,3 +314,69 @@ impl TypeBound {
         }
     }
 }
+
+pub fn expand_type_ref<'a>(
+    db: &dyn DefDatabase,
+    module_id: ModuleId,
+    type_ref: &'a TypeRef,
+) -> Option<Cow<'a, TypeRef>> {
+    let macro_call = match type_ref {
+        TypeRef::Macro(macro_call) => macro_call,
+        _ => return Some(Cow::Borrowed(type_ref)),
+    };
+
+    let file_id = macro_call.file_id;
+    let macro_call = macro_call.to_node(db.upcast());
+
+    let mut expander = Expander::new(db, file_id, module_id);
+    let expanded = expand(db, &mut expander, &macro_call, true)?;
+
+    let node = ast::Type::cast(expanded)?;
+
+    let ctx = LowerCtx::new(db, file_id);
+    return Some(Cow::Owned(TypeRef::from_ast(&ctx, node)));
+
+    fn expand(
+        db: &dyn DefDatabase,
+        expander: &mut Expander,
+        macro_call: &ast::MacroCall,
+        expect_type: bool,
+    ) -> Option<SyntaxNode> {
+        let (mark, mut expanded) = match expander.enter_expand_raw(db, macro_call.clone()) {
+            Ok(ExpandResult { value: Some((mark, expanded)), .. }) => (mark, expanded),
+            _ => return None,
+        };
+
+        if expect_type && !ast::Type::can_cast(expanded.kind()) {
+            expander.exit(db, mark);
+            return None;
+        }
+
+        if ast::MacroType::can_cast(expanded.kind()) {
+            expanded = expanded.first_child()?; // MACRO_CALL
+        }
+
+        let mut rewriter = SyntaxRewriter::default();
+
+        let children = expanded.descendants().filter_map(ast::MacroCall::cast);
+        for child in children {
```

Comment:
this part is surprising to me. I wouldn't expect this function to try to recursively expand all macro calls inside the expansion, and that seems to be what's happening here? I'm not that familiar with the macro expansion code though.

Reply:
This path was actually a remnant of the approach before calling this expansion code in `TypeRef -> Ty` lowering. However, I thought perhaps it would've been good for us to further recurse here so the recusion limit still bounds the recursion depth. But even without recursion, this would have been called "recursively" by the `TypeRef -> Ty` lowering. What do you think? Should I remove this and just let the lowering deal with this?

Reply:
Yes, I think this should be done by the lowering code.

Reply:
On the other hand, if we do it in `hir_def` we don't have to make `Expander` public and deal with it in `TypeLoweringCtx`. Maybe for a future refactor.

Reply:
Dealing with the `Expander` in `TypeLoweringCtx` seems like the correct thing to do for me. I think this syntax-based "expand all macro call descendants" would likely have subtle bugs? E.g. consider something like a macro type expanding to `[u8; some_macro_call!()]`. Or even `[u8; { macro_rules! inner_macro {...}; inner_macro!() }]`...